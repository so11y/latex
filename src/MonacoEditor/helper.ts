import { Node } from "acorn"

export interface ValidateTypeResult {
  node: Node,
  through: boolean,
  message: string | null
}

export function throwNode(node: Node): ValidateTypeResult {
  return {
    node,
    through: true,
    message: null
  }
}

export function extractTokenAndNumbers(str: string) {
  // 匹配 "(1:22)" 格式的模式
  const regex = /\((\d+):(\d+)\)/;

  // 使用正则表达式进行匹配
  const match = str.match(regex);

  // 如果匹配成功，则返回提取的内容，包括两个数字
  if (match) {
    const extracted = {
      token: match[0], // 完整匹配的字符串
      firstNumber: parseInt(match[1]), // 第一个数字
      secondNumber: parseInt(match[2]) // 第二个数字
    };
    return extracted;
  } else {
    return null;
  }
}
