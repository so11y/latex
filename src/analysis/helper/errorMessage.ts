export function formatterError(_: TemplateStringsArray, ...arg: any[]) {
  const [message, ...temp] = arg;
  return temp.reduce((prev, next, index) => {
    return prev.replace(`$${index}`, next);
  }, message);
}

export const ErrorMessage = {
  Unknown: {
    UnKnownCall: "未知方法调用",
    UnknownSyntax: "未知语法",
    UnknownIdentifier: "未知字符",
    UnknownOperator: "不支持位运算等符号",
  },
  BinaryExpression: {
    EqNumberOrCall: "运算两边需要是数值或者函数调用表达式或者运算",
  },
  CallExpression: {
    ArgLengthExpect: "参数个数不匹配需要 $0 个参数",
  },
  ConditionalExpression: {
    OnlyLogical: `只能是使用三元和逻辑表达式符号 $0`,
    OnlyTrueCall: "真结果需要是函数调用并且不能再返回逻辑表达式了",
    OnlyFalseCall: "假结果需要是函数调用并且不能再返回逻辑表达式了",
    NotNestConditionalExpression:
      "不支持在直接嵌套Conditional函数和三元表达式,但是可以在搭配逻辑表达式使用",
  },
  Expression: {
    RootNotBool: "当前最终运算得出的可能是一个布尔值，请检查",
    NotCanRootAst: "最外层只能是函数调用或者运算或者三元表达在或者是数值",
  },
  Identifier: {
    CallKeyCode: "需要调用函数的方式写$0(word)",
  },
  Literal: {
    OnlyInCall: "字符串只能出现在调用里",
  },
  LogicalExpression: {
    OnlyLogical: "逻辑表达式两边需要是条件表达式 使用 && 或者 || 连接 ",
  },
  NumberLiteral: {
    NeedComposeBinary: "数字需要与运算符号配合",
  },
};
