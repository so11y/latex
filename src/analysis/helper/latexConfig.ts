import { ValidateGuardFalseMate } from "../types";
import { StringAccept } from "../callAccept/defaultAccept";

export interface ConfigRecord {
  [key: string]: LatexCallConfigType;
}

export interface LatexValidateCallAccept {
  key: string;
  describe: string;
  validate: (
    this: Omit<LatexCallConfigType["config"], "accept">,
    node: any,
    parent: any,
    index: number
  ) => true | ValidateGuardFalseMate;
}

export interface LatexCallConfigType {
  name: string;
  alias: string;
  astName?: string;
  config: {
    accept: Array<LatexValidateCallAccept>;
  };
}

export class LatexConfig {
  LatexCallConfig = {
    Count: cratedDefaultAccept("Count", "数据量"),
    Mean: cratedDefaultAccept("Mean", "平均值"),
    Median: cratedDefaultAccept("Median", "中位数"),
    Absolute: cratedDefaultAccept("Absolute", "绝对值"),
    Min: cratedDefaultAccept("Min", "最小值"),
    Max: cratedDefaultAccept("Max", "最大值"),
    First: cratedDefaultAccept("First", "第一个点"),
    Last: cratedDefaultAccept("Last", "最后一个点"),
  } satisfies ConfigRecord;

  get LatexNames() {
    return Object.values(this.AllLatex).map((latex) => latex.name);
  }

  get AllLatex() {
    return {
      ...this.LatexCallConfig,
      ...macroLatexCallConfig,
    };
  }
}

export const macroLatexCallConfig = {
  Conditional: {
    name: "Conditional",
    alias: "三元表达式",
  },
};

function cratedDefaultAccept(name: string, alias: string) {
  return {
    name,
    alias,
    config: {
      accept: [StringAccept],
    },
  };
}
