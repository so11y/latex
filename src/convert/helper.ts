import { omit } from "lodash-es";
import { ContainLatexComponentStore, ignoredPropsKey } from "../components/types";
import { StoreContainProvide } from "../components/storeContain";

export function buildConfig(v: ContainLatexComponentStore) {
  const ignorePackOptionKeys = [
    ...ignoredPropsKey,
    "children",
    "hasChildrenBobbleColumn",
    "isLast",
    "isFiled",
    "dept",
  ]
  if (v.data && Object.keys(v.data).length === 0) {
    ignorePackOptionKeys.push("data");
  }
  const _props = omit(v, ignorePackOptionKeys);
  return _props;
}

export function encodeBracket(v: ContainLatexComponentStore, s: StoreContainProvide) {
  const [first] = v.children.value;
  const nest = s.getComponent(first.id);
  nest.bracket = buildConfig(v);
  return nest;
}
