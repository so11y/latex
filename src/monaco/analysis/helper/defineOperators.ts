import { AstType } from "../types";
import { isLogicalOperators } from "../util/functional";

export const operators = {
  "+": {
    latexOperator: "+",
    name: AstType.BinaryExpression,
    alias: "Plus",
  },
  "-": {
    latexOperator: "-",
    name: AstType.BinaryExpression,
    alias: "Minus",
  },
  "*": {
    latexOperator: "*",
    name: AstType.BinaryExpression,
    alias: "Times",
  },
  "/": {
    latexOperator: "{\\div}",
    name: AstType.BinaryExpression,
    alias: "Div",
  },
  "==": {
    latexOperator: "\\equiv",
    name: AstType.LogicalExpression,
    alias: "Eq",
  },
  "!=": {
    latexOperator: "\\ne",
    name: AstType.LogicalExpression,
    alias: "NEq",
  },
  ">": {
    latexOperator: ">",
    name: AstType.LogicalExpression,
    alias: "Gt",
  },
  "<": {
    latexOperator: "<",
    name: AstType.LogicalExpression,
    alias: "Lt",
  },
  ">=": {
    latexOperator: "\\ge",
    name: AstType.LogicalExpression,
    alias: "EGt",
  },
  "<=": {
    latexOperator: "\\ge",
    name: AstType.LogicalExpression,
    alias: "ELt",
  },
  "&&": {
    latexOperator: "\\&\\&",
    name: AstType.LogicalExpression,
    alias: "And",
  },
  "||": {
    latexOperator: "\\|",
    name: AstType.LogicalExpression,
    alias: "Or",
  },
};

export const LogicalOperators =
  Object.keys(operators).filter(isLogicalOperators);
