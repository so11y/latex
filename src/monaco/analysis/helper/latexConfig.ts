import { ValidateSchemaGuardMate } from "../types";
import { DefaultAccept } from "../callAccept/defaultAccept";

interface ConfigRecord {
  [key: string]: LatexCallConfig;
}

export interface LatexValidateConfig {
  validate: (
    this: Omit<LatexCallConfig["config"], "accept">,
    node: any,
    parent: any,
    index: number
  ) => true | ValidateSchemaGuardMate;
}

export interface LatexCallConfig {
  name: string;
  alias: string;
  astName?: string;
  config: {
    accept: Array<LatexValidateConfig>;
  };
}

export const LatexCallConfig = {
  Count: cratedDefaultAccept("Count", "数据量"),
  Mean: cratedDefaultAccept("Mean", "平均值"),
  Median: cratedDefaultAccept("Median", "中位数"),
  Absolute: cratedDefaultAccept("Absolute", "绝对值"),
  Min: cratedDefaultAccept("Min", "最小值"),
  Max: cratedDefaultAccept("Max", "最大值"),
  First: cratedDefaultAccept("First", "第一个点"),
  Last: cratedDefaultAccept("Last", "最后一个点"),
} satisfies ConfigRecord;

export const macroLatexCallConfig = {
  Conditional: {
    name: "Conditional",
    alias: "三元表达式",
  },
};

export const LatexNames = Object.keys({
  ...LatexCallConfig,
  ...macroLatexCallConfig,
});

export const AllLatex = {
  ...LatexCallConfig,
  ...macroLatexCallConfig,
};

function cratedDefaultAccept(name: string, alias: string) {
  return {
    name,
    alias,
    config: {
      accept: DefaultAccept,
    },
  };
}
