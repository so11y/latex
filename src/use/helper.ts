import { LatexComponent } from "../components/types"
import { ComponentName } from "./constName";


export function inspectType(l: LatexComponent, name: ComponentName) {
  return l.componentName === name || l.latexName === name
}
