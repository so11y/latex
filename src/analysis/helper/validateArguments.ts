import { Node } from "estree";
import { LatexValidateAccept } from "./latexConfig";
import { ValidateGuardFalseMate, cratedFalseThrough } from "../types";
import { ErrorMessage, formatterError } from "./errorMessage";
import { LatexSyntax } from "../analysisAst";
import { ForkLatex, Latex } from "../latex";
import { buildTypeNames } from "../util/functional";

export function validateArguments(
  this: Latex,
  n: Array<Node>,
  parent: Node,
  accept: Array<LatexValidateAccept>,
  len: number
) {
  let _this = this;
  if (n.length !== len && len !== Infinity) {
    return cratedFalseThrough(
      parent,
      formatterError`${ErrorMessage.CallExpression.ArgLengthExpect} ${len}`
    );
  }
  //标准化accept参数
  function nomadizeAccept(node: Node, index: number) {
    return {
      node,
      accept: len === Infinity ? accept[0] : accept[index],
    };
  }
  //验证accept参数，如果不通过返回错误信息
  function nomadizeValidateResult(
    _config: ReturnType<typeof nomadizeAccept>,
    index: number
  ) {
    const test = _config.accept.validate.call(
      _this,
      _config.node,
      parent,
      index
    );
    if (test !== true) {
      return cratedFalseThrough(_config.node, test.message!);
    }
    return {
      ..._config,
      through: true,
    };
  }
  //过滤错误的节点
  function filterErrorNodes(config: ReturnType<typeof nomadizeValidateResult>) {
    return config.through === false;
  }

  const nomadizeAccepts = n.map(nomadizeAccept);

  const errorNodes = nomadizeAccepts
    .map(nomadizeValidateResult)
    .filter(filterErrorNodes) as Array<ValidateGuardFalseMate>;
  if (errorNodes.length) {
    return cratedFalseThrough(
      parent,
      null,
      errorNodes.map(({ node, message }) => {
        return cratedFalseThrough(node, message!);
      })
    );
  }

  let childAccept = nomadizeAccepts
    .filter((config) => config.accept.fork)
    .map(({ accept, node }) => {
      const latexSyntax = LatexSyntax();
      const syntaxDefine = accept.fork!.map((key) => {
        return (latexSyntax as any)[`${key}Define`];
      });
      const fork = new ForkLatex(buildTypeNames(syntaxDefine));
      fork.currentAcceptConfig = accept.child!;
      return fork.validate(null, node as any);
    })
    .map(({ diagnosisNodes }) => diagnosisNodes)
    .flat(Infinity) as Array<ValidateGuardFalseMate>;

  if (childAccept.length) {
    return cratedFalseThrough(parent, null, childAccept);
  }

  return true;
}
