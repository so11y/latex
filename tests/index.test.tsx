import { describe, expect, test, } from "vitest"
import { createSimpleApp } from "./helper/createApp"
import { NButton } from "naive-ui"
import StoreContain, { StoreContainProvide } from "../src/components/storeContain"
import { nextTick } from "vue"
import { DOMWrapper } from "@vue/test-utils"


test("mount latex", async () => {
  expect.assertions(2);
  let store: StoreContainProvide
  const app = createSimpleApp((s) => {
    store = s;
    return <>
      <NButton onClick={
        () => s.putActiveChildren('Number', '数值', 'NumberContain')
      } />
    </>
  });
  const input =
    app.getComponent(
      StoreContain
    ).get('[data-id="-1"] > input');

  (input.element as HTMLInputElement).click();
  //@ts-ignore
  if (store) {
    expect(store.getActiveId()).toBe("-1")

    store.putActiveChildren(
      'Count', '数据量', 'SingleContain'
    )
    const current_id = store.getIncreaseUid(true);
    await nextTick();
    store.setActiveId(current_id);

    const count_input = app.get(`[data-id="${current_id}"] > span > input`);

    (count_input.element as HTMLInputElement).click()

    await nextTick()
    const dialog_input = document.querySelector(".n-input__input-el") as HTMLInputElement;
    await nextTick();
    const wraper = new DOMWrapper(dialog_input)

    wraper.setValue("xixi")
    // wraper.element.value

    const dialog_buttons = document.querySelector(".n-dialog__action");
    (dialog_buttons?.children.item(1) as HTMLButtonElement).click();

    await nextTick();

    expect(app.html()).toMatchSnapshot();
  }
})
