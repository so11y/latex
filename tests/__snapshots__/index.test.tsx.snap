// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mount latex 1`] = `
"<!--teleport start-->
<!--teleport end-->
<button class="n-button n-button--default-type n-button--medium-type" tabindex="0" type="button" style="--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-bezier-ease-out: cubic-bezier(0, 0, .2, 1); --n-ripple-duration: .6s; --n-opacity-disabled: 0.5; --n-wave-opacity: 0.6; font-weight: 400; --n-color: #0000; --n-color-hover: #0000; --n-color-pressed: #0000; --n-color-focus: #0000; --n-color-disabled: #0000; --n-ripple-color: #18a058; --n-text-color: rgb(51, 54, 57); --n-text-color-hover: #36ad6a; --n-text-color-pressed: #0c7a43; --n-text-color-focus: #36ad6a; --n-text-color-disabled: rgb(51, 54, 57); --n-border: 1px solid rgb(224, 224, 230); --n-border-hover: 1px solid #36ad6a; --n-border-pressed: 1px solid #0c7a43; --n-border-focus: 1px solid #36ad6a; --n-border-disabled: 1px solid rgb(224, 224, 230); --n-width: initial; --n-height: 34px; --n-font-size: 14px; --n-padding: 0 14px; --n-icon-size: 18px; --n-icon-margin: 6px; --n-border-radius: 3px;">
  <!---->
  <transition-stub name="fade-in-width-expand-transition" appear="false" persisted="false" css="true">
    <!---->
  </transition-stub>
  <!---->
  <div aria-hidden="true" class="n-base-wave"></div>
  <div aria-hidden="true" class="n-button__border"></div>
  <div aria-hidden="true" class="n-button__state-border"></div>
</button>
<div data-v-73788115="" class="latex-contain" data-id="-1">
  <div data-v-73788115="" style="color: #CC99FF;" class="latex-contain__title">
    <div data-v-73788115="" class="remove">x</div> Root<span data-v-73788115="" class="bracket">(</span>
  </div>
  <div data-v-73788115="" class="latex_gap">
    <div data-v-73788115="" class="latex-contain" data-id="1">
      <div data-v-73788115="" style="color: #66CCFF;" class="latex-contain__title">
        <div data-v-73788115="" class="remove">x</div> Count<span data-v-73788115="" class="bracket">(</span>
      </div>
      <div data-v-73788115="" class="latex_gap">
        <div data-v-73788115="" class="latex-contain" data-id="2">
          <div data-v-73788115="" style="color: #003366;" class="latex-contain__title">
            <div data-v-73788115="" class="remove">x</div> xixi
            <!--v-if-->
          </div>
          <!--v-if-->
        </div>
      </div><span><!--v-if--></span>
      <div data-v-73788115="" style="color: #66CCFF;" class="bracket bracket-end"> ) </div>
    </div>
  </div>
  <!--v-if-->
  <div data-v-73788115="" style="color: #CC99FF;" class="bracket bracket-end"> ) </div>
</div>"
`;
